### PR TITLE
Simplify order lines bookkeeping.

### PR DESCRIPTION
**Pull request by fpacanowski**

### CODE REVIEW

File: ecommerce/ordering/lib/ordering/order.rb [hunk 1]
```diff
@@ -79,35 +79,34 @@ def cancel
...
-      return unless order_line
-      order_line.decrease_quantity
-      remove_order_line(order_line) if order_line.empty?
+      @basket.decrease_quantity(event.data[:product_id])
+    end
+  end
+
+  class Basket
+    def initialize
+      @order_lines = {}
```
Comment (jandudulski) [inline ecommerce/ordering/lib/ordering/order.rb:92]: How about:

```suggestion
      @order_lines = Hash.new { |basket, product_id| basket[product_id] = 0 }
```

and you don't need `fetch(..., 0)` anymore

File: ecommerce/ordering/lib/ordering/order.rb [hunk 2]
```diff
@@ -79,35 +79,34 @@ def cancel
...
 
-    private
+    def increase_quantity(product_id)
+      @order_lines[product_id] = quantity(product_id) + 1
+    end
 
-    def find_order_line(product_id)
-      @order_lines.select { |line| line.product_id == product_id }.first
+    def decrease_quantity(product_id)
+      return unless @order_lines.has_key?(product_id)
```
Comment (jandudulski) [inline ecommerce/ordering/lib/ordering/order.rb:100]: How about such case:

```ruby
basket.increase_quantity(id)
2.times { basket.decrease_quantity(id) }
```

File: ecommerce/ordering/lib/ordering/order.rb [hunk 3]
```diff
@@ -79,35 +79,34 @@ def cancel
...
-      product_id = event.data[:product_id]
-      order_line = find_order_line(product_id)
-      return unless order_line
-      order_line.decrease_quantity
-      remove_order_line(order_line) if order_line.empty?
+      @basket.decrease_quantity(event.data[:product_id])
+    end
+  end
+
+  class Basket
```
Comment (jandudulski) [inline ecommerce/ordering/lib/ordering/order.rb:90]: How about extracting to another file?

**Comment by andrzejkrzywda** [comment]
Thanks for the improvement @fpacanowski.

I like this class extraction. I'm not sure about the name (Order consists of Basket?), but it feels OK for now.

Thanks for the comments @jandudulski. I'm not sure about extracting this class to another file in this case.

Anyway, I gave you both Write accesses, feel free to improve the codebase and push directly to master, good luck :)

**Comment by fpacanowski** [comment]
Thanks. I addressed your comments in https://github.com/RailsEventStore/ecommerce/commit/787213adf04b449418f5431f12d8ea9d520674ab.

**Pull request merged**